kinode/src/fd_manager.rs

```diff
@@ -0,0 +1,274 @@
+use lib::types::core::{
+    Address, FdManagerError, FdManagerRequest, KernelMessage, Message, MessageReceiver,
+    MessageSender, PrintSender, Printout, ProcessId, Request, FD_MANAGER_PROCESS_ID,
+};
+use std::{collections::HashMap, sync::Arc};
+
+const DEFAULT_MAX_OPEN_FDS: u64 = 180;
+const DEFAULT_FDS_AS_FRACTION_OF_ULIMIT_PERCENTAGE: u64 = 60;
+const DEFAULT_UPDATE_ULIMIT_SECS: u64 = 3600;
+const DEFAULT_CULL_FRACTION_DENOMINATOR: u64 = 2;
+
+struct State {
+    fds: HashMap<ProcessId, u64>,
+    mode: Mode,
+    total_fds: u64,
+    max_fds: u64,
+    cull_fraction_denominator: u64,
+}
+
+enum Mode {
+    /// don't update the max_fds except by user input
+    StaticMax,
+    /// check the system's ulimit periodically and update max_fds accordingly
+    DynamicMax {
+        max_fds_as_fraction_of_ulimit_percentage: u64,
+        update_ulimit_secs: u64,
+    },
+}
+
+impl State {
+    fn new() -> Self {
+        Self::default()
+    }
+
+    fn default() -> Self {
+        Self {
+            fds: HashMap::new(),
+            mode: Mode::default(),
+            total_fds: 0,
+            max_fds: DEFAULT_MAX_OPEN_FDS,
```

Comment:
Shouldn't `update_max_fds` be called once here at the start, so default limit is not applied briefly at launch?